db/db_iter.cc

```diff
@@ -152,8 +152,12 @@ void DBIter::Next() {
 
   local_stats_.next_count_++;
   if (ok && iter_.Valid()) {
-    FindNextUserEntry(true /* skipping the current user key */,
-                      prefix_same_as_start_);
+    if (prefix_same_as_start_ && prefix_extractor_ != nullptr) {
```

Comment:
In DBIter, can we set prefix_same_as_start_ to false if prefix_extractor_ is nullptr? This should avoid lots of conjunction in the code.

Reply:
Good idea. We can do it in the constructor. Let me do it in a separate pull request.